### PR TITLE
Refuse to upgrade database on worker processes

### LINKED ISSUE #8006: restarting multiple workers at once risks applying database migrations multiple times
each worker independently checks if the schema is up-to-date and applies the migrations if not. At best this fails with exceptions; at worst it could result in data corruption

### PR DESCRIPTION
**Pull request by richvdh**
Fixes #8006.

I'm not 100% convinced this is what we want to do (aren't we trying to get rid of the master process?) but for now this is the best thing I can think of to do.

### CODE REVIEW

File: synapse/storage/prepare_database.py [hunk 1]
```diff
@@ -71,25 +71,36 @@ def prepare_database(db_conn, database_engine, config, databases=["main", "state
...
-                    databases=databases,
+                raise ValueError(
+                    "config==None in prepare_database, but databse is not empty"
+                )
+
+            # if it's a worker app, refuse to upgrade the database, to avoid multiple
+            # workers doing it at once.
+            if config.worker_app is not None and user_version != SCHEMA_VERSION:
+                raise UpgradeDatabaseException(
+                    "Expected database schema version %i but got %i"
```
Comment (erikjohnston) [inline synapse/storage/prepare_database.py:85]: Do we want to update these error messages to say that master should be started first? I believe these are fairly user visible errors and so it'd be good to be clear how to fix the problem.

**Comment by erikjohnston** [comment]
> I'm not 100% convinced this is what we want to do (aren't we trying to get rid of the master process?) but for now this is the best thing I can think of to do.

We could do the same thing as the media repository where we specify which instance runs the updates in the config. But that feels like overkill for now

File: synapse/storage/prepare_database.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/prepare_database.py]: fair. done, I hope.

**Review by erikjohnston — APPROVED** [review]
Woot, thanks!

**Pull request merged**